### PR TITLE
feat: introduce bag-of-constraints solver

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
The diff is very large, it is just because of a missing merge/rebase? I feel like I still dont understand git.

**Comment by mlutze** [comment]
> The diff is very large, it is just because of a missing merge/rebase? I feel like I still dont understand git.

Yeah, the git ghost is about. I'll fix it sometime.

**Review by magnus-madsen — COMMENTED** [review]
One thing:

How about doing changes to TypeReduction and MapOps in two separate PRs?

**Comment by magnus-madsen** [comment]
Can you write me a short description of how to best review this?

**Comment by mlutze** [comment]
> Can you write me a short description of how to best review this?

No, but I can write a long one.

### Before this PR

Unification of types is handled in two different places:
* `ConstraintSolver` is called by `Typer`
* `Unification` is called everywhere else: `Monomorph`, `HoleCompletion`, `EffectVerifier`, `Instances`, `Stratifier`, as well as `ConstraintSolver` itself.
* `Unification` is just a wrapper around `OldStarUnification`

### This PR

This PR does 2 things:
- introduces `ConstraintSolver2`, a bag-of-constraints, phase-based parser
-  changes `Unification` to call `ConstraintSolver2` instead of `OldStarUnification`

Other changes are administrative:
- Copy of `RecordConstraintSolver` using `ConstraintSolver2` interface
- ditto `SchemaConstraintSolver`
- ditto `TypeReduction`
- Cutting off `RecordUnification` and `SchemaUnification` from the new `Unification`

#### `ConstraintSolver2`

This solver applies the unification rules as phases. The high-level algorithm is:
```python
while progress:
    for each phase:
        for each constraint:
            apply the phase's rule to the constraint
```

Each phase is represented by a function taking a type constraint and returning one of the following:
- a type constraint
- a set of type constraints
- a set of type constraints and a _substitution tree_

These phases are applied to a collection of constraints through an abstraction `Soup`, which allows a pipeline-like syntax for applying rules, while hiding management of substitutions.

The substitution tree is not used in this PR, as it is intended for Purification constraints, which do not arise from calling `Unification` functions. You can think of the substitution tree as a simple substitution for now.

### After this PR

At least three PRs will follow.

1. A PR to remove code made redundant by this PR. This includes:
  - `RecordUnification`
  - `SchemaUnification`
  - `EqualityEnvironment`
  - `AssocTypeSubstitution`
  - `OldStarUnification`
  - possibly more
2. A PR to replace `ConstraintSolver` with `ConstraintSolver2`. This will likely reveal some bugs.
3. A PR to remove code made redundant by PR 2. (By overwriting it with `ContraintSolver2`'s helpers. This includes:
  - `TypeReduction`
  - `RecordConstraintSolver`
  - `SchemaConstraintSolver`
  - probably more

**Comment by magnus-madsen** [comment]
In the final version, how will Monomorpher conduct unification? 

Will it use constraints? One constraint? Or will it just call deeper into the solver (i.e. some specific unification procedure?)

**Comment by mlutze** [comment]
> In the final version, how will Monomorpher conduct unification? 
> 
> Will it use constraints? One constraint? Or will it just call deeper into the solver (i.e. some specific unification procedure?)

It will use one constraint. That behavior is introduced by this PR, through an interface.

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import ca.uwaterloo.flix.language.phase.typer.ConstraintSolver2.{Progress, TypeConstraint}
+import ca.uwaterloo.flix.language.phase.unification.Substitution
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+object RecordConstraintSolver2 {
+
+  /**
+    * Unifies the two given record row types.
+    */
+  def solve(tpe1: Type, tpe2: Type)(implicit scope: Scope, tracker: Progress, renv: RigidityEnv, flix: Flix): (List[TypeConstraint], Substitution) = (tpe1, tpe2) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:30]: I would consider putting the tracker as the last argument just before flix.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:30]: What do you think of making scope and renv explicit arguments? It seems they are *actual* arguments to the function.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:30]: tracker: Progress -- So which is it? :)

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    // ----------
+    // ρ ~ ρ => ∅
+    case (t1, t2) if t1 == t2 =>
+      tracker.markProgress()
+      (Nil, Substitution.empty)
+
+    //    α ∉ fv(ρ)
+    // ----------------
+    // α ~ ρ  =>  {α ↦ ρ}
+    case (Type.Var(sym, _), tpe) if !tpe.typeVars.exists(_.sym == sym) && renv.isFlexible(sym) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: Perhaps we need to bite the bullet and add `contains`?

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    case Type.Var(sym, loc) if !tvars.contains(sym) && renv.isFlexible(sym) =>
+      val tvar = Type.freshVar(Kind.RecordRow, loc)
+      val newRow = Type.mkRecordRowExtend(hdLabel, hdTpe, tvar, loc)
+      val subst = Substitution.singleton(sym, newRow)
+      Some((newRow, subst))
+
+    // If no rule matches, then we cannot pivot this row type.
+    case _ => None
+  }
+}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:125]: I must say I think this is pretty clean. Great job 👍

File: main/src/ca/uwaterloo/flix/language/phase/typer/SchemaConstraintSolver2.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{Kind, Name, RigidityEnv, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.typer.ConstraintSolver2.{Progress, TypeConstraint}
+import ca.uwaterloo.flix.language.phase.unification.Substitution
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+//
+// This is a copy of the RecordConstraintSolver2. Because the shapes are slightly different, we cannot reuse the same solver.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/SchemaConstraintSolver2.scala:26]: See comments from other file, I suppose.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+import org.apache.commons.lang3.reflect.{ConstructorUtils, MethodUtils}
+
+import java.lang.reflect.{Constructor, Field, Method}
+
+object TypeReduction2 {
+
+  /**
+    * Performs various reduction rules on the given type.
+    */
+  def reduce(tpe0: Type)(implicit scope: Scope, tracker: Progress, renv0: RigidityEnv, eqenv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef], flix: Flix): Type = tpe0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:34]: same comment about actual arguments vs. other stuff.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    case t: Type.Var => t
+
+    case t: Type.Cst => t
+
+    case Type.Apply(tpe1, tpe2, loc) =>
+      // TODO CONSTR-SOLVER-2 this is recursive. Might be a hot-spot for performance?
+      val t1 = reduce(tpe1)
+      val t2 = reduce(tpe2)
+      Type.Apply(t1, t2, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:43]: if (t1 eq tpe1) .... etc

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+    case Type.Alias(cst, args, tpe, loc) => tpe
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:45]: Unused args?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 5 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:34]: eqenv -> eqenv0?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+      matches match {
+        // Case 1: No match. Can't reduce the type.
+        case Nil => tpe0
+
+        // Case 2: One match. Use it.
+        case newTpe :: Nil =>
+          tracker.markProgress()
+          newTpe
+
+        // Case 3: Multiple matches. Give back the original type.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:76]: How can this happen?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        // TODO CONSTR-SOLVER-2 Right resiliency strategy?
+        case _ :: _ :: _ => tpe0
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:78]: No error?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          Type.getFlixType(method.getReturnType)
+
+        case Type.Cst(TypeConstructor.JvmField(field), _) =>
+          tracker.markProgress()
+          Type.getFlixType(field.getType)
+
+        case t => Type.JvmToType(t, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:95]: Reuse tpe0 and case on _?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+      reduce(tpe) match {
+        case Type.Cst(TypeConstructor.JvmConstructor(constructor), _) =>
+          tracker.markProgress()
+          BaseEffects.getConstructorEffs(constructor, loc)
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          BaseEffects.getMethodEffs(method, loc)
+
+        case t => Type.JvmToType(t, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:108]: Reuse tpe0 and case on _?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+      }
+
+    case unresolved@Type.UnresolvedJvmType(member, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:111]: But unresolved is tpe0!

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+            case JavaMethodResolution.Resolved(method) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmMethod(method), loc)
+            case _ => unresolved
+          }
+      }
+  }
+
+  /** Tries to find a constructor of `clazz` that takes arguments of type `ts`. */
+  def lookupConstructor(clazz: Class[?], ts: List[Type]): JavaConstructorResolution = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:148]: Is the following standard / reused? Do I need to check it?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+    // Check if we found a matching constructor.
+    if (c != null) {
+      JavaConstructorResolution.Resolved(c)
+    } else {
+      JavaConstructorResolution.NotFound
+    }
+  }
+
+  /** Tries to find a method of `thisObj` that takes arguments of type `ts`. */
+  def lookupMethod(thisObj: Type, methodName: String, ts: List[Type]): JavaMethodResolution = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:164]: What methods should be public vs private?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+
+    /** One matching field. */
+    case class Resolved(field: Field) extends JavaFieldResolution
+
+    /** No matching field. */
+    case object NotFound extends JavaFieldResolution
+
+    /**
+      * The types of the lookup are not resolved enough to decide
+      * (it contains type variables, associated types, etc.).
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:308]: it -> they
drop parens
Fix grammar and formatting.

Same comment applies to JavaMethodRes

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+  object JavaMethodResolution {
+
+    /** One matching method. */
+    case class Resolved(method: Method) extends JavaMethodResolution
+
+    /** No matching method. */
+    case object NotFound extends JavaMethodResolution
+
+    /**
+      * The types of the lookup are not resolved enough to decide
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:326]: see above

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+  object JavaConstructorResolution {
+
+    /** One matching constructor. */
+    case class Resolved(constructor: Constructor[?]) extends JavaConstructorResolution
+
+    /** No matching constructor. */
+    case object NotFound extends JavaConstructorResolution
+
+    /**
+      * The types of the lookup are not resolved enough to decide
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:345]: see above

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+      * (they contain type variables, associated types, etc.).
+      */
+    case object UnresolvedTypes extends JavaMethodResolution
+
+  }
+
+  /** A lookup result of a Java constructor. */
+  sealed trait JavaConstructorResolution
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:334]: Move JavaConstructorResolution before JavaMethodResolution

**Comment by magnus-madsen** [comment]
I still need to look at ConstraintSolver2

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala [hunk 19]
```diff
@@ -0,0 +1,580 @@
...
+import ca.uwaterloo.flix.util.collection.{ListMap, MapOps}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Result}
+
+/**
+  * The constraint solver reduces a collection of constraints by iteratively applying reduction rules.
+  * The result of constraint solving is a substitution and a list of constraints that could not be resolved.
+  */
+object ConstraintSolver2 {
+
+  sealed trait TypeConstraint
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala:33]: What do you say to moving TypeConstraint, Progress, Soup, SubstitutionTree into their own files?

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: You mean Type.contains?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 5 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:34]: I use `0` to mark when a variable can change in a recursive call. So actually it should be `renv` :)

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: Yes

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: Thoughts?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: Actually, does tpe.typeVars.contains not work?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 9 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:78]: This solver doesn't know about errors.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 13 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:148]: All of the `lookup*` stuff is copy-paste

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 14 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:164]: Probably almost all private

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala (hunk 19 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala:33]: Yes

**Comment by magnus-madsen** [comment]
Adding some source comments for some of my comments might be useful.

File: main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/RecordConstraintSolver2.scala:41]: It works, but I am against it on principle :)

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala (hunk 6 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala:43]: I'll do this in a follow-up.

**Comment by mlutze** [comment]
> Adding some source comments for some of my comments might be useful.

I do not know what this means.

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala (hunk 19 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala:33]: Except Soup, which is linked to what goes on here.

**Comment by magnus-madsen** [comment]
I will review this soon again, but I need Jonathans PR to land first. So you should also harass him :)

**Comment by magnus-madsen** [comment]
@mlutze There are no conflicts. Does this mean I can merge this or do you want to experiment?

**Comment by mlutze** [comment]
> @mlutze There are no conflicts. Does this mean I can merge this or do you want to experiment?

Let's at least wait for one tick on the arewefast graph :)

**Comment by JonathanStarup** [comment]
Will this PR not use `SetUnification.unifyAll`? or is that later?

**Comment by mlutze** [comment]
> Will this PR not use `SetUnification.unifyAll`? or is that later?

Later. This one only deals with one constraint at a time anyway.

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@mlutze You can merge this tomorrow.

**Pull request merged**